### PR TITLE
feat(rust): incremental index update

### PR DESCRIPTION
**Pull request by eddyxu**
`Dataset::optimize_index()` will update the index with newly added data, without retrain the IVF index.

### CODE REVIEW

File: rust/lance/src/index/vector/ivf.rs [hunk 1]
```diff
@@ -139,6 +156,59 @@ impl IVFIndex {
...
+
+        let pq_index = self
+            .sub_index
+            .as_any()
+            .downcast_ref::<PQIndex>()
+            .ok_or(Error::Index {
+                message: "Only support append to IVF_PQ".to_string(),
+            })?;
+
+        // TODO: merge two IVF implementations.
```
Comment (westonpace) [inline rust/lance/src/index/vector/ivf.rs:183]: I'm not sure I understand this TODO

File: rust/lance/src/index/vector/ivf/io.rs [hunk 2]
```diff
@@ -28,29 +33,45 @@ pub(super) async fn write_index_partitions(
...
+            let part = existing_idx.load_partition(part_id as usize).await?;
+            let pq_idx = part.as_any().downcast_ref::<PQIndex>().unwrap();
+            if pq_idx.code.is_some() {
+                let pq_code_arr = pq_idx.code.as_ref().unwrap().clone();
+                let pq_code_fixed_size_arr = FixedSizeListArray::try_new_from_values(
+                    pq_code_arr.as_ref().clone(),
+                    pq_idx.num_sub_vectors as i32,
+                )?;
+                pq_array.push(Arc::new(pq_code_fixed_size_arr));
+                println!("pq_idx.row_ids: {:#?}", pq_idx.row_ids.as_ref());
```
Comment (westonpace) [inline rust/lance/src/index/vector/ivf/io.rs:52]:
```suggestion
```
Or change to a log

**Review by westonpace — APPROVED** [review]
This looks very clean.  It might be nice to have at least one end-to-end python test.  I can add one if you want (it should be similar to the compaction one)

**Comment by changhiskhan** [comment]
We should also test combinations of operations. Like index then some combinations of {add, delete, merge, compact} then incremental index ?

**Comment by eddyxu** [comment]
> We should also test combinations of operations. Like index then some combinations of {add, delete, merge, compact} then incremental index ?

Lets do that in python? it is much easier in python

File: rust/lance/src/index/vector/ivf/io.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/io.rs]: Oh this was a left over for debugging. removed it.

File: rust/lance/src/index/vector/ivf.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf.rs:183]: We have half-migrated IVF model to `lance-index`.  I will find sometime next week to move `Ivf` model completely from `lance` crate

**Comment by eddyxu** [comment]
> This looks very clean. It might be nice to have at least one end-to-end python test. I can add one if you want (it should be similar to the compaction one)

Can not say no to the offer :)  @westonpace

**Pull request merged**